### PR TITLE
feat: implement client wrapper around discovery service API

### PR DESCRIPTION
**Pull request by smira**
This wrapper handles protobuf marshaling, encryption, etc. on the client
data so that service doesn't have a way to see the data.

Client handles data refresh on TTL, updates, discovery of other
affiliates, etc.

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

### CODE REVIEW

**Comment by smira** [comment]
TODO: 

- [ ] handle hello
- [ ] implement more aggressive test with more clients and a complete mess of updates

File: api/v1alpha1/client/affiliate.proto [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+message IPPrefix {
+    bytes ip = 1;
+    int32 bits = 2;
+}
+
+// Endpoint for the Affiliate KubeSpan spec as seen internally in the client code.
+//
+// Discovery service receives Affiliate marshaled to protobuf and encrypted.
+message Endpoint {
+    bytes ip = 1;
```
Comment (AlekSi) [inline api/v1alpha1/client/affiliate.proto:37]: For IPv4 address, will that field contain 4 or 7-15 bytes? Maybe we could use `string` type to make that more clear?

If that's too late to change the type everywhere, we probably could add fields comments to explain that.

File: api/v1alpha1/client/affiliate.proto [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    int32 port = 2;
```
Comment (AlekSi) [inline api/v1alpha1/client/affiliate.proto:38]: Could it be negative? If not, could we use uint32?

Comment (smira) [reply api/v1alpha1/client/affiliate.proto:38]: it can't, good idea

File: api/v1alpha1/client/affiliate.proto (hunk 1 above)
Comment (smira) [reply api/v1alpha1/client/affiliate.proto:37]: it uses standard (more or less) binary marshaling for IPs which is implemented in `netaddr` library; I belive Linux kernel uses same at the very least, but good idea to have that documented

string representation is much larger

**Review by Unix4ever — APPROVED** [review]

File: api/v1alpha1/client/affiliate.proto (hunk 1 above)
Comment (Ulexus) [reply api/v1alpha1/client/affiliate.proto:40]: Yeah, I'd rather stick with the binary form

**Review by Ulexus — APPROVED** [review]

**Pull request merged**